tokio/src/sync/mpsc/bounded.rs

```diff
@@ -463,6 +463,39 @@ impl<T> Receiver<T> {
         self.chan.close();
     }
 
+    /// Checks if a channel is closed.
+    ///
+    /// This method returns `true` if the channel has been closed and there are
+    /// no remaining messages in the channel's buffer. The channel is closed
+    /// when all senders have been dropped, or when [`close`] is called.
+    ///
+    /// # Examples
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = mpsc::channel(10);
+    ///     assert!(!rx.is_closed());
+    ///
+    ///     tx.send(100).await.unwrap();
+    ///     tx.send(200).await.unwrap();
+    ///
+    ///     rx.close();
+    ///     assert!(!rx.is_closed());
+    ///     assert_eq!(rx.recv().await, Some(100));
+    ///
+    ///     assert!(!rx.is_closed());
+    ///     assert_eq!(rx.recv().await, Some(200));
+    ///     assert!(rx.recv().await.is_none());
+    ///     
+    ///     assert!(rx.is_closed());
+    /// }
+    /// ```
+    pub fn is_closed(&mut self) -> bool {
```

Comment:
So I was going through this one more time and I realized that you may be right here. Currently, it's possible to have a situation where `Sender::is_closed` returns true but `Receiver::is_closed()` returns false. That's probably not desirable. It probably makes more sense to change it to match the existing method on the sender.

Of course, another option is to rename this method to reflect its actual behavior.

Reply:
> Currently, it's possible to have a situation where Sender::is_closed returns true but Receiver::is_closed() returns false. That's probably not desirable. It probably makes more sense to change it to match the existing method on the sender.

Yeah, I totally agree with this.  I think the `is_closed` should return true if `close` was called or all the senders are dropped, regardless of having any messages available in the channel.

I believe it would be a good idea to provide an additional function, maybe `is_empty` or `has_messages`, that checks if there are any messages available in the channel, this way the user would be able to check both  conditions with `rx.is_closed() && rx.is_empty()`.

What do you think?

Reply:
That seems reasonable to me. Adding an `is_empty` method could make even more sense together with adding a `len` method (which would close #6314). After all, clippy emits a warning if you add `len` without `is_empty`.

But if you only want to add `is_empty`, that's also okay.

Reply:
I have added the `is_empty` and `len` functions. 

The `len` implementation only checks for the `tail_position`. I am not sure if it is necessary to check the `ready` bits for valid values in each position, since that would be relatively costly because of the linked list implementation

Reply:
Hmm, regarding the ready bits, I guess the main question here is concurrency-related. For example, imagine if you have the following two threads running in parallel:
```Rust
sender.send("message1").await;
assert!(receiver.len() != 0); // or similar with `is_empty`
```
```Rust
sender.send("message2").await;
```
Then, depending on how you implement `len`, it's possible that the assertion could fail. This would happen if message1 ends up being stored _after_ message2 in the channel queue, but when the assert runs, message2 is has not yet been fully sent.

We once had a similar bug with `try_recv`. See #1997. Basically, this assert could fail:
```Rust
sender.send("message1").await;
assert!(receiver.try_recv().is_some());
```
```Rust
sender.send("message2").await;
```
The assert would sometimes fail when message2 ends up before message1 in the queue. Here, `try_recv` can't return message1 because we guarantee that messages are returned in order. It also can't return message2 because the call to `send("message2")` is still running. So it returns `None` instead, even though we _know_ that a message has been successfully sent. When we fixed it in #4113, we did so by having `try_recv` sleep until `send("message2")` finished when there are fully sent messages later in the queue.

To test this, you can add these loom tests:
```Rust
// tokio/src/sync/tests/loom_mpsc.rs
#[test]
fn len_nonzero_after_send() {
    loom::model(|| {
        let (send, recv) = mpsc::unbounded_channel();
	let send2 = send.clone();

	let join = thread::spawn(move || {
	    block_on(send2.send("message2")).unwrap();
	});

        block_on(send.send("message1")).unwrap();
        assert!(recv.len() != 0);

        join.join().unwrap();
    });
}

#[test]
fn nonempty_after_send() {
    loom::model(|| {
        let (send, recv) = mpsc::unbounded_channel();
	let send2 = send.clone();

	let join = thread::spawn(move || {
	    block_on(send2.send("message2")).unwrap();
	});

        block_on(send.send("message1")).unwrap();
        assert!(!recv.is_empty());

        join.join().unwrap();
    });
}
```
I don't know whether these tests will fail, but if they do, please look at `try_recv` to see how it determines whether it should sleep or not.

Reply:
Both tests failed. I am looking on how to fix it

Reply:
I added both loom tests and fixed the `is_empty` implementation.

About `len_nonzero_after_send` test, I think the `len` implementation was already correct, but I ran the wrong test initially. 
At first I had written the assertion `assert!(recv.len() == 2)`, which should fail since the `join()` is after the `assert!`.

With the `assert!(recv.len() != 0);` assertion, the test passes